Add unwrap phase functionality

po09i opened the pull request:
This PR is to set up the initial unwrapping structure as well as an unwrapping algorithm (prelude). 

The main purposes of this PRs:
- Add example script: examples/unwrap_script.py
- Implemented one of the uwrapping algorithm: prelude: shimmingtoolbox/unwrap/prelude.py (wrapper)
- Implement tests

Fixes #10 
Fixes #17 
Fixes #23

Aureliengilliot commented:
Tried implementing the fitting procedure on the signal with complete periods, but it fails because the signal is unbalanced:
![untitled](https://user-images.githubusercontent.com/48367658/54393573-e4ce6300-4680-11e9-9fa0-db5a7a18ab93.png)

Next thing to try is to average the signal within period, and to fit that.

jcohenadad commented:
OK, for now i've opted for an approach where the poly degree is updated after X timepoints. This is very empirical but should work OK for now. We can improve it in the future.

jcohenadad commented:
@rtopfer @alexfoias please let us know how this function should be called by the rest of the code, so we can finalized i/o stuff

rtopfer commented:
It should be called (and possibly embedded as a subfunction) within [ProbeTracking.getupdate()](https://github.com/neuropoly/realtime_shimming/blob/84821815ebdf9ba0c171660900b3263c51b0d6f4/RespiratoryProbe/ProbeTracking.m#L191)

Nicolas-Pinon commented:
Simple filtering did not give better result than the polynomial fitting (frequency tuning of the filter was a pain)
To overcome the problem of the low frequency part of the signal varying too much, the fitting of the polynomial curve is now only updated **every period** of the signal instead of every time point, this way there is less small variations.
This also make sense because there is no point in updating the fit every point since we are not interested in the high frequency variations for the fit.

jcohenadad commented:
@Nicolas-Pinon thank you for the investigations. Would it be possible to point to a dataset and procedure for testing (in order to review this PR). Thanks!

Nicolas-Pinon commented:
> @Nicolas-Pinon thank you for the investigations. Would it be possible to point to a dataset and procedure for testing (in order to review this PR). Thanks!

Right now I will be going to my brass band rehearsal, could that be a work for @Aureliengilliot ?
I'll help him if there is trouble, but by looking at the before/after my fix, I'm pretty confident in it.

jcohenadad commented:
> looking at the before/after my fix, I'm pretty confident in it

documenting a testing framework will enable to derive unit testing in the future, and make sure the algorithm will still work until the years 2050

alexfoias commented:
@Nicolas-Pinon We have tested the implementation in real time. The current version (fitting based on each period) doesn't seem to work. 

You can test the algorithm with this dataset. 
[20190606_11-33-03AM_VariablesDump_au.mat.zip](https://github.com/neuropoly/realtime_shimming/files/3262421/20190606_11-33-03AM_VariablesDump_au.mat.zip)

jcohenadad commented:
guys, before moving forward, [as I already asked](https://github.com/neuropoly/realtime_shimming/pull/15#issuecomment-498839493) please document how the testing is done (if each of you test in a different way, we are not making progress). So, on top of this PR, document:
- what command to run in order to run the function with a specified mat file
- list a couple of datasets used for validating this PR

when you test and report in this PR, always indicate the git version so we can compare across previous implementations

alexfoias commented:
I documented more the previous test on the initial comment.

alexfoias commented:
This comment: https://github.com/neuropoly/realtime_shimming/pull/15#issuecomment-499559093

jcohenadad commented:
@alexfoias can you add an optional input param to the function `get_capacitive_probe_signal()` so we don't have to modify the code itself for debugging, but instead do something like: `get_capacitive_probe_signal(file_recording=blablabla.mat)`

then, please, update the [top comment](https://github.com/neuropoly/realtime_shimming/pull/15#issue-260869037) to explain how to test the stuff and list a couple of mat files to download for testing.

alexfoias commented:
Implemented the feature here: https://github.com/neuropoly/realtime_shimming/commit/d37d330a1d0ced95e637bbbf42293706ed0796ea

jcohenadad commented:
@rtopfer you mentioned you worked on this files recently to implement the "watchdog feature" (i.e. acquiring from one station, and creating the recording from another station). What branch are these features implemented in?  I'd like to make sure this PR is based on the latest code

rtopfer commented:
@jcohenadad the daemon recording functionality using [ memmapfile()](https://www.mathworks.com/help/matlab/ref/memmapfile.html) is on the master branch. The polynomial fitting, filtering, etc. would ideally be implemented within [ProbeTracking.processupdate()](https://github.com/neuropoly/realtime_shimming/blob/86b198fe9ba24b79569b2f0d81d8f17b2c227f3d/RespiratoryProbe/ProbeTracking.m#L278)

jcohenadad commented:
thank you @rtopfer. This PR needs to be updated then. Also, I strongly recommend to create an external function which would only deal with the detrending, and which would be called [there](https://github.com/neuropoly/realtime_shimming/blob/86b198fe9ba24b79569b2f0d81d8f17b2c227f3d/RespiratoryProbe/ProbeTracking.m#L302) (for code clarity).

alexfoias commented:
@Aureliengilliot @jcohenadad I added the baseline measurement [here](https://github.com/neuropoly/realtime_shimming/tree/alfoi/CapacitiveProbe/test)

jcohenadad commented on RespiratoryProbe/ProbeTracking.m:
```diff
@@ -51,16 +51,24 @@
+    %   https://stackoverflow.com/questions/29671482/private-constructor-in-matlab-oop
+        Aux.launchrecordingdaemon() ; % runs continuously in background
+        return;
+    else
```
@rtopfer when launching `ProbeTracking(recording.mat)`, `Aux` does exists and hence it does not go inside this else. What did you have in mind here?

rtopfer replied:
ProbeTracking() receives an input filepath (leading to .mat) in 2 foreseeable cases:
1. this debugging feature we're implementing now, in which a user directly calls ProbeTracking( _filepath_ ) ;
2. when ProbeTracking() is called by the daemon matlab session in the dual-session mode. That also involves a .mat file, but in the first case that .mat shouldn't contain a variable called 'Aux' (and therefore, exist('Aux') would return false). Clearly that would be violated if for some reason the user decided to name another variable 'Aux'...  I will add other check conditions.

jcohenadad replied:
ok, while you do that, could you please also comment the file header (i've started doing it, but since my understanding of this code is limited, you'd better update my comment)

jcohenadad commented on RespiratoryProbe/ProbeTracking.m:
```diff
@@ -1,22 +1,22 @@
 % =========================================================================
-%
-% =========================================================================
-% Updated::20190528::ryan.topfer@polymtl.ca
```
no date-- GH is there for that

jcohenadad commented on RespiratoryProbe/ProbeTracking.m:
```diff
@@ -51,16 +51,24 @@
+        return;
+    else
+        Specs.state = 'inert' ;
+        error('TODO: enable debug mode: previous (raw) recording is loaded and can be processed as if it were a new recording...')
```
@rtopfer i am not familiar with the deamon procedure (and there is no explanation in your code as to how to set this up)-- so any tips here would be appreciated (what function to call? etc.)

rtopfer replied:
As it stands, ProbeTracking.declareprobe() looks for the device ; if it finds it, the daemon session is launched. (i.e. there is no function to call). 
For the debugging case, we don't need a daemon session (I was suggesting Specs.state = 'inert' as a sort of flag to avoid launching a 2nd session, but this flag could be coded differently (e.g. Specs.state = 'debug'))...

jcohenadad replied:
ok, but the `error()` should ultimately be replaced with "something";  this "something" being what is typically launched by the deamon, so any tips about the workflow of the code would be welcome (could be in the header of the file)

p.s. `Specs.state = 'debug'` sounds better (unless there are other reasons than debugging for it to be inert)

rtopfer commented:
@jcohenadad disabling daemon mode:

_Specs.isRecordingDaemonEnabled = false ;
P = ProbeTracking( Specs ) ;_

Note that _P.recordphysiosignal( )_ in this mode will always begin an entirely new recording (i.e. fitting routines start from scratch)

jcohenadad commented:
@rtopfer when stopping and re-running `P.recordphysiosignal`, i get the following:
~~~
>> P.recordphysiosignal
Error using serial/fopen (line 72)
Open failed: OBJ has already been opened.

Error in ProbeTracking/beginrecording (line 189)
    fopen( Aux.Source ) ;

Error in ProbeTracking/recordphysiosignal (line 606)
isRecording = Aux1.beginrecording() ;
~~~
i need to close/reopen matlab. Any clue how to make it easier?

po09i committed:
NEW : Added unwrap folder structure

po09i committed:
Added numpy requirement

po09i committed:
Remove version from __init__ subfolder

po09i committed:
Initial function for unwrap phase

po09i committed:
NEW : Initial prelude wrapper

po09i committed:
NEW : fill init

po09i committed:
update requirements

po09i committed:
NEW : add debug script

po09i committed:
make script callable from CL

po09i committed:
Fix delete of files

jcohenadad committed:
Added .idea

jcohenadad committed:
Added download of example data

jcohenadad committed:
Added phase loading and diff

po09i committed:
Update script to use test data

po09i committed:
Return 3d instead of 2d

po09i committed:
Added support for 5d complex arrays

po09i committed:
Added better error handling

po09i committed:
Added 2d support

po09i committed:
Updated according to comments

po09i committed:
Update todo for phaseunwrapping

po09i committed:
Added docstring

po09i committed:
moved mask to be defined in script

po09i committed:
moved mask to script

po09i committed:
added docstring

po09i committed:
added todo

po09i committed:
Update docstring

po09i committed:
Default to no mask

po09i committed:
Changed download function

po09i committed:
Removed unnecessary comment

po09i committed:
Resolve issues in script according to comments

po09i committed:
Update according to comments

po09i committed:
Removed interp function

po09i committed:
Added simple test

jcohenadad committed:
Merge branch 'master' into ad/unwrap-phase

jcohenadad committed:
Added deps

jcohenadad committed:
clustered phase code together

jcohenadad committed:
Minor fixes

po09i committed:
Commented out the whole function (not useable anymore)

po09i committed:
Rewrote prelude to have mag and phase inputs

po09i committed:
Removed complex array handling

po09i committed:
Wrote test to accomodate new prelude inputs

po09i committed:
Fix conflicts

po09i committed:
Added download in travis

po09i committed:
forgot a comma

po09i committed:
and another one

po09i committed:
Removed reference to unwrap_phase

po09i committed:
set up tests with the new download function

po09i committed:
Added handling for non default paths

po09i committed:
Added test for non default path

po09i committed:
Added mask tests

po09i committed:
Added tests for mag and phase input

po09i committed:
Added more checks for non_default_masks

jcohenadad committed:
Refactored test to remove duplication

jcohenadad committed:
Now possible to download non-compressed data, e.g. prelude

jcohenadad committed:
Download prelude binary

Related to https://github.com/shimming-toolbox/shimming-toolbox-py/issues/23

po09i committed:
Added click as a dependency

po09i committed:
Fixed hardcoded name when deleting unwrapped nii file

po09i committed:
Calls prelude in tmp folder for every test

po09i committed:
Added default back

po09i committed:
Test prelude -h

po09i committed:
Changed test to .yml file

po09i committed:
Added prelude to the path to be called within script

po09i committed:
Add permissions

po09i committed:
extra char

po09i committed:
Add FSL required variable

po09i committed:
Reduced permissions

po09i committed:
Clean up

po09i committed:
Use download function

jcohenadad committed:
Updated default for mask input argument

jcohenadad committed:
Added API: add_suffix()

jcohenadad committed:
Refactored output of prelude() API

po09i committed:
Removed phase difference todo and opened issue #36

po09i committed:
Updates variables to reflect convention

po09i committed:
Removed error handling from example script

po09i committed:
Resolved comments

po09i committed:
Added 2D as an input type and changed is_unwrapping_in_2d to false

po09i committed:
Removed logging

po09i committed:
Updated tests

po09i committed:
Removed B0 mapping code: out of scope

po09i committed:
Add threshold for more precise masking

po09i committed:
forgot shape

po09i committed:
Apply suggestions from code review

Co-authored-by: Julien Cohen-Adad <jcohen@polymtl.ca>

po09i committed:
Added run_subprocess function

po09i committed:
Use run_subprocess() function

jcohenadad committed:
Merge branch 'master' into ad/unwrap-phase

jcohenadad committed:
Replaced glob.glob by __dir_testing__

jcohenadad committed:
Changed matplotlib for oriented-object use

So it doesn't crash on headless systems, see: https://github.com/neuropoly/spinalcordtoolbox/wiki/Moving-away-from-pyplot

jcohenadad committed:
Cosmetic fixes

jcohenadad committed:
Removed unused file

This file is now referenced in #45

jcohenadad committed:
Merge branch 'master' into ad/unwrap-phase
